File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.go
func minDays(bloomDay []int, m int, k int) int {
	if m*k > len(bloomDay) {
		return -1
	}
	left, right := slices.Min(bloomDay)-1, slices.Max(bloomDay)+1

	isOk := func(day int) bool {
		flower, bouquets := 0, 0
		for i := 0; i < len(bloomDay); i++ {
			if bloomDay[i] <= day {
				flower++
				if flower == k {
					bouquets++
					flower = 0
				}
			} else {
				flower = 0
			}
			if bouquets >= m {
				return true
			}
		}
		return false
	}
	
	for left+1 < right {
		mid := (left + right) / 2
		if isOk(mid) {
			right = mid
		} else {
			left = mid
		}
	}
	return right
}

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.rs
impl Solution {
    pub fn min_days(bloom_day: Vec<i32>, m: i32, k: i32) -> i32 {
        if (bloom_day.len() as i32) < m * k || m * k < 0{
            return -1;
        }

        let max_days = *(bloom_day.iter().max().unwrap());
        let (mut l, mut r) = (-1, max_days+1);

        let check = |days: i32| {
            let mut f = 0;
            let mut bouquet = 0;
            for &i in &bloom_day {
                if days >= i {
                    f += 1;
                } else {
                    f = 0;
                }
                
                if f == k {
                    bouquet += 1;
                    f = 0;
                }
                if bouquet >= m {
                    break;
                }
            }
            bouquet >= m
        };

        while l + 1 != r {
            let mid = (r + l) / 2;
            if check(mid) {
                r = mid
            } else {
                l = mid
            }
        }

        r
    }
}